### DISCUSSION TITLE
R-universe "global" API

### DISCUSSION THREAD

**Discussion by vgherard** [r-universe-org/help#84] [Ideas]
Dear R-universe developers,

thanks for the great service you are providing. 

I'm developing a small R package to interact with R-universes' APIs. By now it correctly implements GET features such as retrieving packages' informations and stats from specific universes - while I'm leaving all features involving POST/PUT/DELETE verbs for a future version, maybe.

I was wondering whether it would be possible, at present, to include also some "global" (i.e. at the level of the whole R-universe.dev) feature; to clarify, some example of global features might be:

1. Retrieving the list of all available universes 
2. Retrieving the list of all available packages (irrespective of universe)
3. Retrieving global statistics on package dependencies
4. ...

For instance, I have currently implemented the first feature by GETting `r-universe.dev/stats/organizations`, whose response is a JSON with the required information. I haven't seen this nor any other "global" API features publicly documented.

I wanted to kindly ask for some feedback in this respect: would you discourage the usage of GET requests to `r-universe.dev` such as the above? Have you already implemented some of the "global" features I described above (and, if yes, is there any place to which I can refer for documentation)?

**Comment by jeroen** [comment]
The API you found shows the global universes and packages: https://r-universe.dev/stats/organizations
There is also one that is grouped by maintainer: https://r-universe.dev/stats/maintainers
The checks are in https://r-universe.dev/stats/checks?limit=100 but they are limited by default because this is a lot of data.

It depends on what you want to do with it. For now you can use them, but the main issue I foresee is that if R universe would become much larger, these APIs may become too heavy and we may have to change them.

**Reply by vgherard** [reply]
Got it, thanks. That's what I thought (hope your foresight turns out right :-) ).

To me, a very simple reason for querying `r-universe.dev/stats/organizations` was that a query such as, *e.g.*, `<username>.r-universe.dev/packages`, returns a response with status 200 even if `<username>` is not an existing universe - such responses may be confused with those from an existing but empty universe - right?

Anyway, I guess I will keep using these few features and stay tuned if things change in future.

Thanks

**Reply by jeroen** [reply]
Technically there is no difference between an empty universe and non-existing user. The server only stores packages and their owner. So a query with zero results is still a valid query.

**Reply by wlandau** [reply]
Is there a way to query specific packages? I'm just looking for an overall pass/fail for each package for https://github.com/r-releases/help/issues/6.
